free_fleet/src/manager/Manager.cpp

```diff
@@ -0,0 +1,435 @@
+auto Manager::robot_names() -> std::vector<std::string>
+{
+  std::lock_guard<std::mutex> lock(_pimpl->mutex);
+  std::vector<std::string> result;
+  result.reserve(_pimpl->robots.size());
+  for (const auto it : _pimpl->robots)
+  {
+    result.push_back(it.first);
+  }
+  return result;
+}
+
+//==============================================================================
+auto Manager::robot(const std::string& robot_name)
+-> std::shared_ptr<const manager::RobotInfo>
+{
+  std::lock_guard<std::mutex> lock(_pimpl->mutex);
+  const auto it = _pimpl->robots.find(robot_name);
+  if (it == _pimpl->robots.end())
+    return nullptr;
+  return it->second;
+}
+
+//==============================================================================
+auto Manager::all_robots()
+-> std::vector<std::shared_ptr<const manager::RobotInfo>>
+{
+  std::lock_guard<std::mutex> lock(_pimpl->mutex);
+  std::vector<std::shared_ptr<const manager::RobotInfo>> infos;
+  infos.reserve(_pimpl->robots.size());
+  for (const auto it : _pimpl->robots)
+  {
+    infos.push_back(it.second);
+  }
+  return infos;
+}
+
+//==============================================================================
+auto Manager::request_pause(const std::string& robot_name)
+-> std::optional<TaskId>
+{
+  std::lock_guard<std::mutex> lock(_pimpl->mutex);
+  if (_pimpl->robots.find(robot_name) == _pimpl->robots.end())
+    return std::nullopt;
+
+  const TaskId request_task_id = ++_pimpl->current_task_id;
+  messages::PauseRequest request(robot_name, request_task_id);
+
+  std::shared_ptr<manager::RequestInfo> request_info(
+    new manager::SimpleRequestInfo<messages::PauseRequest>(
+      request,
+      [this](const messages::PauseRequest& request_msg)
+      {
+        this->_pimpl->middleware->send_pause_request(request_msg);
+      },
+      [this]() {return _pimpl->time_now_fn();}));
+
+  _pimpl->tasks[request_task_id] = request_info;
+  _pimpl->unacknowledged_tasks[request_task_id] = request_info;
+  request_info->send_request();
+  return _pimpl->current_task_id;
+}
+
+//==============================================================================
+auto Manager::request_resume(const std::string& robot_name)
+-> std::optional<TaskId>
+{
+  std::lock_guard<std::mutex> lock(_pimpl->mutex);
+  if (_pimpl->robots.find(robot_name) == _pimpl->robots.end())
+    return std::nullopt;
+
+  const TaskId request_task_id = ++_pimpl->current_task_id;
+  messages::ResumeRequest request(robot_name, request_task_id);
+
+  std::shared_ptr<manager::RequestInfo> request_info(
+    new manager::SimpleRequestInfo<messages::ResumeRequest>(
+      request,
+      [this](const messages::ResumeRequest& request_msg)
+      {
+        this->_pimpl->middleware->send_resume_request(request_msg);
+      },
+      [this]() {return _pimpl->time_now_fn();}));
+
+  _pimpl->tasks[request_task_id] = request_info;
+  _pimpl->unacknowledged_tasks[request_task_id] = request_info;
```

Comment:
For each robot, any new new task (really, command) should supersede previous tasks (really, commands). But in the current implementation, the robot will be forced to acknowledge every single request that gets sent.

As an alternative approach, I might recommend mapping unacknowledged tasks/commands with the robot ID as the key, and only keeping up to one unacknowledged task at a time (the one with the highest task ID value) per robot. I'd also recommend wrapping that check-and-overwrite into a simple member function that each of these `request_...` functions can easily call.